### PR TITLE
[Seq] Add an inner symbol to permit references to clock gates

### PR DESCRIPTION
**Pull request by nandor**
The inner symbol of a clock gate translates to an inner symbol on the instance it lowers to. Also slightly adjusted the lowering to cache constants and produce more readable test output.

### CODE REVIEW

File: lib/Dialect/Seq/Transforms/ExternalizeClockGate.cpp [hunk 1]
```diff
@@ -61,43 +62,54 @@ void ExternalizeClockGatePass::runOnOperation() {
...
-    // a `comb.or` to merge the two enable conditions.
-    if (hasTestEnable && !testEnable)
-      testEnable = builder.create<ConstantOp>(i1Type, 0);
-
-    // If the clock gate has no test enable operand but the module does, add a
-    // constant 0 input.
-    if (!hasTestEnable && testEnable) {
-      enable = builder.createOrFold<comb::OrOp>(enable, testEnable, true);
-      testEnable = {};
+  for (auto &[module, clockGatesToReplace] : gatesInModule) {
```
Comment (dtzSiFive) [inline lib/Dialect/Seq/Transforms/ExternalizeClockGate.cpp:72]: Looks like `module` is unused, what's the reason for processing these in groups by module (re-use of the constant)?

File: lib/Dialect/Seq/Transforms/ExternalizeClockGate.cpp [hunk 2]
```diff
@@ -61,43 +62,54 @@ void ExternalizeClockGatePass::runOnOperation() {
...
+    Value cstFalse;
+    for (auto ckgOp : clockGatesToReplace) {
+      ImplicitLocOpBuilder builder(ckgOp.getLoc(), ckgOp);
+
+      Value enable = ckgOp.getEnable();
+      Value testEnable = ckgOp.getTestEnable();
+
+      // If the clock gate has a test enable operand but the module does not,
+      // add a `comb.or` to merge the two enable conditions.
```
Comment (dtzSiFive) [inline lib/Dialect/Seq/Transforms/ExternalizeClockGate.cpp:81]: Pre-existing, but.. aren't the comments above this and next block backwards?

File: lib/Dialect/Seq/Transforms/ExternalizeClockGate.cpp [hunk 3]
```diff
@@ -61,43 +62,54 @@ void ExternalizeClockGatePass::runOnOperation() {
...
+
+      instPorts.push_back(ckgOp.getInput());
+      instPorts.push_back(enable);
+      if (testEnable)
+        instPorts.push_back(testEnable);
+
+      auto instOp = builder.create<InstanceOp>(
+          externModuleOp, builder.getStringAttr(instName), instPorts);
+
+      if (auto sym = ckgOp.getInnerSymAttr())
```
Comment (dtzSiFive) [inline lib/Dialect/Seq/Transforms/ExternalizeClockGate.cpp:104]: If you'd like, could collapse this into the `InstanceOp` creation, although would need to provide empty `parameters`.

**Review by dtzSiFive — APPROVED** [review]
LGTM, re:op + inner symbol usage.

File: lib/Dialect/Seq/Transforms/ExternalizeClockGate.cpp (hunk 1 above)
Comment (nandor) [reply lib/Dialect/Seq/Transforms/ExternalizeClockGate.cpp:72]: Tests are annoying to write and debug when they are littered with %false_0, %false_100, etc.

**Comment by fabianschuiki** [comment]
Thanks for adding this :+1:

**Pull request merged**